File: backend/src/services/settlementService.js
'use strict';

const { knex } = require('../db/knex');
const { withTransaction } = require('../db/tx');
const settlementRepo = require('../repositories/settlementRepo');
const assessmentRepo = require('../repositories/assessmentRepo');
const claimRepo = require('../repositories/claimRepo');
const communications = require('./communicationService');
const audit = require('./auditService');
const claimStatus = require('../domain/claimStatus');
const settlement = require('../domain/settlement');
const { withStatusView } = require('./claimService');
const { dec, toAmount } = require('../utils/money');
const { NotFoundError, ValidationError, ConflictError } = require('../utils/errors');
const logger = require('../config/logger');

async function forClaim(claimId) {
  return settlementRepo.findByClaim(knex, claimId);
}

/** Settlement screen model: claim + settlement + payment tracker + suggested amount. */
async function settlementScreen(claimId) {
  const claim = await claimRepo.findById(knex, claimId);
  if (!claim) return null;
  const [s, assessment] = await Promise.all([
    settlementRepo.findByClaim(knex, claimId),
    assessmentRepo.findByClaim(knex, claimId),
  ]);
  let suggested;
  if (s && s.finalAmount != null) suggested = s.finalAmount;
  else if (assessment && assessment.netPayable != null) suggested = assessment.netPayable;
  else suggested = claim.estimatedLoss;
  return {
    claim: withStatusView(claim),
    settlement: s,
    tracker: settlement.TRACKER,
    suggestedAmount: toAmount(suggested),
    paymentMethods: settlement.PAYMENT_METHODS,
  };
}

/** Authorise (or re-amount) a settlement. Requires an approved/processing claim. */
async function authorize(actor, claimId, amount, fields, ip) {
  if (amount == null || dec(amount).lessThanOrEqualTo(0)) {
    throw new ValidationError('Enter a valid settlement amount.');
  }
  const amt = toAmount(amount);
  return withTransaction(async (trx) => {
    const claim = await claimRepo.findById(trx, claimId);
    if (!claim) throw new NotFoundError('Claim not found.');
    const existing = await settlementRepo.findByClaim(trx, claimId);
    if (!existing) {
      if (claim.status !== claimStatus.APPROVED) {
        throw new ConflictError('Only approved claims can be settled.');
      }
      await settlementRepo.insert(trx, {
        claimId,
        finalAmount: amt,
        justification: fields.justification,
        paymentMethod: fields.paymentMethod,
        accountHolder: fields.accountHolder,
        bankName: fields.bankName,
        accountNumber: fields.accountNumber,
        ifscCode: fields.ifscCode,
        status: settlement.AUTHORIZED,
        approvedBy: Number(actor.id),
      });
      await claimRepo.updateStatus(trx, claimId, claimStatus.SETTLEMENT_PROCESSING);
      await communications.system(
        trx,
        claimId,
        'ICMS',
        `Settlement of ₹ ${amt} has been authorised for your claim.`
      );
    } else {
      await settlementRepo.updateAmount(trx, claimId, amt, fields.justification);
    }
    await writeAudit(trx, actor, 'SETTLEMENT_AUTHORIZED', `${claim.claimNo} (₹${amt})`, ip);
    logger.info({ actor: actor.username, amount: amt, claimNo: claim.claimNo }, 'Settlement authorised');
  });
}

/** Advance the payment tracker one step; returns the new status (or current if at end). */
async function advance(actor, claimId, ip) {
  return withTransaction(async (trx) => {
    const claim = await claimRepo.findById(trx, claimId);
    const s = await settlementRepo.findByClaim(trx, claimId);
    if (!claim || !s) throw new NotFoundError('No settlement to advance.');
    const nextStatus = settlement.next(s.status);
    if (!nextStatus) return s.status; // already closed / unknown

    await settlementRepo.advance(trx, claimId, nextStatus);

    if (nextStatus === settlement.PAYMENT_CONFIRMED) {
      await claimRepo.updateStatus(trx, claimId, claimStatus.SETTLED);
      await communications.system(
        trx,
        claimId,
        'ICMS',
        `Your claim has been settled. Payment of ₹ ${s.finalAmount} is confirmed.`
      );
    } else if (nextStatus === settlement.CLOSED) {
      await claimRepo.updateStatus(trx, claimId, claimStatus.CLOSED);
    }
    await writeAudit(trx, actor, `SETTLEMENT_${nextStatus}`, claim.claimNo, ip);
    return nextStatus;
  });
}

async function writeAudit(trx, actor, action, entity, ip) {
  return audit.record(trx, {
    userId: Number(actor.id),
    username: actor.username,
    role: actor.role,
    action,
    entity,
    ipAddress: ip,
    result: 'SUCCESS',
  });
}

module.exports = { forClaim, settlementScreen, authorize, advance };

File: src/main/java/com/sgx/icms/service/SettlementService.java
package com.sgx.icms.service;

import java.math.BigDecimal;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.sgx.icms.dao.ClaimDao;
import com.sgx.icms.dao.SettlementDao;
import com.sgx.icms.db.Db;
import com.sgx.icms.domain.AuditLog;
import com.sgx.icms.domain.Claim;
import com.sgx.icms.domain.ClaimStatus;
import com.sgx.icms.domain.Settlement;
import com.sgx.icms.web.support.SessionUser;

/**
 * Settlement authorisation and payment tracking. Authorising creates the settlement
 * and moves the claim into SETTLEMENT_PROCESSING; advancing walks the payment tracker
 * and syncs the claim to SETTLED / CLOSED at the right points.
 */
public class SettlementService {

    private static final Logger LOG = LoggerFactory.getLogger(SettlementService.class);

    private final SettlementDao settlementDao = new SettlementDao();
    private final ClaimDao claimDao = new ClaimDao();
    private final CommunicationService communications = new CommunicationService();
    private final AuditService audit = new AuditService();

    public Settlement forClaim(long claimId) {
        return Db.withConnection(conn -> settlementDao.findByClaim(conn, claimId));
    }

    /** Authorise (or re-amount) a settlement. Requires an approved/processing claim. */
    public void authorize(SessionUser actor, long claimId, BigDecimal amount, String paymentMethod,
                          String accountHolder, String bankName, String accountNumber, String ifsc,
                          String justification, String ip) {
        if (amount == null || amount.signum() <= 0) {
            throw new IllegalArgumentException("Enter a valid settlement amount.");
        }
        Db.inTransaction(conn -> {
            Claim claim = claimDao.findById(conn, claimId);
            if (claim == null) {
                throw new IllegalStateException("Claim not found.");
            }
            Settlement existing = settlementDao.findByClaim(conn, claimId);
            if (existing == null) {
                if (!ClaimStatus.APPROVED.equals(claim.getStatus())) {
                    throw new IllegalStateException("Only approved claims can be settled.");
                }
                Settlement s = new Settlement();
                s.setClaimId(claimId);
                s.setFinalAmount(amount);
                s.setJustification(justification);
                s.setPaymentMethod(paymentMethod);
                s.setAccountHolder(accountHolder);
                s.setBankName(bankName);
                s.setAccountNumber(accountNumber);
                s.setIfscCode(ifsc);
                s.setStatus(Settlement.AUTHORIZED);
                s.setApprovedBy(actor.getId());
                settlementDao.insert(conn, s);
                claimDao.updateStatus(conn, claimId, ClaimStatus.SETTLEMENT_PROCESSING);
                communications.system(conn, claimId, "ICMS",
                        "Settlement of ₹ " + amount + " has been authorised for your claim.");
            } else {
                settlementDao.updateAmount(conn, claimId, amount, justification);
            }
            writeAudit(conn, actor, "SETTLEMENT_AUTHORIZED", claim.getClaimNo() + " (₹" + amount + ")", ip);
            LOG.info("{} authorised settlement {} for claim {}", actor.getUsername(), amount, claim.getClaimNo());
            return null;
        });
    }

    /** Advance the payment tracker one step; returns the new status (or current if at end). */
    public String advance(SessionUser actor, long claimId, String ip) {
        return Db.inTransaction(conn -> {
            Claim claim = claimDao.findById(conn, claimId);
            Settlement s = settlementDao.findByClaim(conn, claimId);
            if (claim == null || s == null) {
                throw new IllegalStateException("No settlement to advance.");
            }
            int idx = Settlement.TRACKER.indexOf(s.getStatus());
            if (idx < 0 || idx >= Settlement.TRACKER.size() - 1) {
                return s.getStatus(); // already closed / unknown
            }
            String next = Settlement.TRACKER.get(idx + 1);
            settlementDao.advance(conn, claimId, next);

            if (Settlement.PAYMENT_CONFIRMED.equals(next)) {
                claimDao.updateStatus(conn, claimId, ClaimStatus.SETTLED);
                communications.system(conn, claimId, "ICMS",
                        "Your claim has been settled. Payment of ₹ " + s.getFinalAmount() + " is confirmed.");
            } else if (Settlement.CLOSED.equals(next)) {
                claimDao.updateStatus(conn, claimId, ClaimStatus.CLOSED);
            }
            writeAudit(conn, actor, "SETTLEMENT_" + next, claim.getClaimNo(), ip);
            return next;
        });
    }

    private void writeAudit(java.sql.Connection conn, SessionUser actor, String action, String entity, String ip) {
        AuditLog a = new AuditLog();
        a.setUserId(actor.getId());
        a.setUsername(actor.getUsername());
        a.setRole(actor.getRole());
        a.setAction(action);
        a.setEntity(entity);
        a.setIpAddress(ip);
        a.setResult(AuditLog.RESULT_SUCCESS);
        audit.record(conn, a);
    }
}
